### ISSUE TITLE
Rename `StateScoped` to `StateExitDespawn`, add the inverse variant `StateEntryDespawn`

### ISSUE THREAD

**Issue by Jaso333** [bevyengine/bevy#15849]
## What problem does this solve or what need does it fill?

The current naming of the `StateScoped` component is slightly misleading: it could be interpreted as "this entity will not exist outside of this state". Whereas in fact it merely despawns on exit of the given state, it doesn't care if you spawn an entity _outside_ of the given state:

```
for (entity, binding) in &query {
    if binding.0 == *exited {
        #[cfg(feature = "bevy_hierarchy")]
        commands.entity(entity).despawn_recursive();
        #[cfg(not(feature = "bevy_hierarchy"))]
        commands.entity(entity).despawn();
    }
}
```

To me, this functionality is fine, I can't think of a situation where you would want to spawn something in a state it doesn't belong, only for it to be deleted instantly.

There is also no available _inverse_ of this component: despawn an entity on _entry_ to a state, which I recently found a need for, and have written a very similar mechanism to achieve the same thing.

## What solution would you like?

- Rename `StateScoped` to `StateExitDespawn`, or maybe a cleaner looking name than that.
- Add the inverse version: `StateEntryDespawn`, having it operate during state transition, before `OnEnter`.

Here's a snippet of what I produced for my own game:

```
impl AppExtLifetime for SubApp {
    fn enable_state_entry_despawn<S: States>(&mut self) -> &mut Self {
        self.add_systems(
            StateTransition,
            state_entry_despawn::<S>
                .in_set(TransitionSchedules::<S>::default())
                .run_if(on_event::<StateTransitionEvent<S>>()),
        );
        self
    }
}
```
## What alternative(s) have you considered?

No real alternatives.

## Additional context

State transitions are such a powerful and flexible way to orchestrate game entity lifecyles and the flow of the game itself, it would be nice to see more features in this area.

**Comment by Jaso333** [comment]
> IMO having entities "scoped to a particular state" is more simple concept than despawn-on-entry/exit. If `StateScoped` gets warning (or even instant despawn) against insertion during mismatched state, then these new components could be added to fill the gaps for advanced use.

Yes, maybe a warning will help to better specify the intention of `StateScoped` and having the other two I mention as separate.

I would say that the additional ones I'm suggesting are not "advanced" though, it seems more of a necessity as I develop a game driven by states and substates. I'm currently only developing a Pong clone, but rather than just throwing it together, I'm trying to build scalable infrastructure and learn from it going forward. I have a top-level `GameState` which contains `Menu` and `Level`, and then inside the "level" is `Countdown` and `Play` substates. Everything I spawn in the "level" gets scoped to the `Level` game state, so if at any point the player quits back to the menu, all entities in the level are despawned. The round starts with a `Countdown` and then transitions to `Play` when the timer has finished. I spawn the paddles and ball on entry to `Countdown`, which signifies a new round. When play has finished in the `Play` state, I want to loop back round to `Countdown` state to start a new round, and I want to simply respawn the ball and paddles. Therefore, I actually want to scope the paddles and ball to _two_ states, which is where the current options fall short. By having a despawn-on-entry, I'm able to simply transition from `Play` back to `Countdown` and the `StateEntryDespawn` takes care of despawning the previous ball and paddles.

**Comment by benfrankel** [comment]
Note that this also opens the door to expanding the type into an enum as such:

```rust
pub enum DespawnOnExit<S: State> {
    Single,
    #[default]
    Recursive,
    Descendants,
    #[doc(hidden)]
    _PhantomData(PhantomData<S>),
}
```

Also "despawn on enter", "hide on enter", "hide on exit", "disable on enter", etc. become very reasonable to add.
